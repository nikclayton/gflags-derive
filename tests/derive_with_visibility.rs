extern crate gflags_derive;

mod common;
use common::*;

// Derive the flags in an inner module. The generated `TO_STDERR` variable
// should not be visible outside this module, the `DIR` variable should be.
//
// This is one half of this test -- verifying that `DIR` is visible in the
// `super` crate.
//
// The second half is in `expected_failures/derive_with_visibility.rs` to
// verify that trying to access `TO_STDERR` generates a compile-time error.
mod inner_for_test {
    use gflags_derive::GFlags;

    #[derive(GFlags)]
    #[allow(dead_code)]
    struct Config {
        /// True if log messages should also be sent to STDERR
        to_stderr: bool,

        /// The directory to write log files to
        #[gflags(visibility = "pub(super)")]
        dir: String,
    }
}

#[test]
fn derive_with_visibility() {
    let mut flags = fetch_flags();

    check_flag(
        Some(ExpectedFlag::<&str> {
            doc: &["The directory to write log files to"],
            name: "dir",
            placeholder: None,
            generated_flag: &inner_for_test::DIR,
        }),
        flags.remove("dir"),
    );
}
