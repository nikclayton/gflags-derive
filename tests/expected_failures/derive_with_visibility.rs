extern crate gflags_derive;

// Derive the flags in an inner module. The generated `TO_STDERR` variable
// should not be visible outside this module, the `DIR` variable should be.
//
// This is one half of this test -- verifying that `TO_STDERR` is not visible
// in the `super` crate.
//
// The second half is in `../derive_with_visibility.rs` to verify that trying
// to access `DIR` works.
mod inner_for_test {
    use gflags_derive::GFlags;

    #[derive(GFlags)]
    #[allow(dead_code)]
    struct Config {
        /// True if log messages should also be sent to STDERR
        to_stderr: bool,

        /// The directory to write log files to
        #[gflags(visibility = "pub(super)")]
        dir: String,
    }
}

fn main() {
    if inner_for_test::TO_STDERR.is_present() {
        panic!("Can't happen, flag is not visible in this scope");
    }
}
