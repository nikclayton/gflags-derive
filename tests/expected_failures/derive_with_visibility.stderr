error[E0603]: static `TO_STDERR` is private
  --> $DIR/derive_with_visibility.rs:27:24
   |
27 |     if inner_for_test::TO_STDERR.is_present() {
   |                        ^^^^^^^^^ this static is private
   |
note: the static `TO_STDERR` is defined here
  --> $DIR/derive_with_visibility.rs:14:14
   |
14 |     #[derive(GFlags)]
   |              ^^^^^^
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
